include/libunwind-riscv64.h

```diff
@@ -0,0 +1,180 @@
+
+#ifndef LIBUNWIND_H
+#define LIBUNWIND_H
+
+#if defined(__cplusplus) || defined(c_plusplus)
+extern "C" {
+#endif
+
+#include <sys/types.h>
+#include <inttypes.h>
+#include <ucontext.h>
+
+#define UNW_TARGET              riscv64
+#define UNW_TARGET_RISCV64      1
+
+#define _U_TDEP_QP_TRUE 0       /* ignored - see libunwind-dynamic.h */
+
+/* This needs to be big enough to accommodate "struct cursor", while
+   leaving some slack for future expansion.  Changing this value will
+   require recompiling all users of this library.  Stack allocation is
+   relatively cheap and unwind-state copying is relatively rare, so we
+   want to err on making it rather too big than too small.  */
+/* FIXME for riscv64: Figure out a more reasonable size */
+#define UNW_TDEP_CURSOR_LEN     4096
+
+typedef uint64_t unw_word_t;
+typedef int64_t unw_sword_t;
+
+#if defined __riscv_float_abi_double
+// We assume FLEN=64 for the F registers
+typedef double unw_tdep_fpreg_t;
+#else
+# error "Unsupported RISC-V floating point ABI"
+#endif
+
+/* Also see src/riscv64/Gglobal.c. This ordering is consistent with
+   https://github.com/riscv/riscv-elf-psabi-doc/blob/74ecf07bcebd0cb4bf3c39f3f9d96946cd6aba61/riscv-elf.md#dwarf-register-numbers- */
+
+typedef enum
+  {
+    /* integer registers */
+    UNW_RISCV64_X0,
+    UNW_RISCV64_X1,
+    UNW_RISCV64_X2,
+    UNW_RISCV64_X3,
+    UNW_RISCV64_X4,
+    UNW_RISCV64_X5,
+    UNW_RISCV64_X6,
+    UNW_RISCV64_X7,
+    UNW_RISCV64_X8,
+    UNW_RISCV64_X9,
+    UNW_RISCV64_X10,
+    UNW_RISCV64_X11,
+    UNW_RISCV64_X12,
+    UNW_RISCV64_X13,
+    UNW_RISCV64_X14,
+    UNW_RISCV64_X15,
+    UNW_RISCV64_X16,
+    UNW_RISCV64_X17,
+    UNW_RISCV64_X18,
+    UNW_RISCV64_X19,
+    UNW_RISCV64_X20,
+    UNW_RISCV64_X21,
+    UNW_RISCV64_X22,
+    UNW_RISCV64_X23,
+    UNW_RISCV64_X24,
+    UNW_RISCV64_X25,
+    UNW_RISCV64_X26,
+    UNW_RISCV64_X27,
+    UNW_RISCV64_X28,
+    UNW_RISCV64_X29,
+    UNW_RISCV64_X30,
+    UNW_RISCV64_X31,
+
+    /* floating point registers */
+    UNW_RISCV64_F0,
+    UNW_RISCV64_F1,
+    UNW_RISCV64_F2,
+    UNW_RISCV64_F3,
+    UNW_RISCV64_F4,
+    UNW_RISCV64_F5,
+    UNW_RISCV64_F6,
+    UNW_RISCV64_F7,
+    UNW_RISCV64_F8,
+    UNW_RISCV64_F9,
+    UNW_RISCV64_F10,
+    UNW_RISCV64_F11,
+    UNW_RISCV64_F12,
+    UNW_RISCV64_F13,
+    UNW_RISCV64_F14,
+    UNW_RISCV64_F15,
+    UNW_RISCV64_F16,
+    UNW_RISCV64_F17,
+    UNW_RISCV64_F18,
+    UNW_RISCV64_F19,
+    UNW_RISCV64_F20,
+    UNW_RISCV64_F21,
+    UNW_RISCV64_F22,
+    UNW_RISCV64_F23,
+    UNW_RISCV64_F24,
+    UNW_RISCV64_F25,
+    UNW_RISCV64_F26,
+    UNW_RISCV64_F27,
+    UNW_RISCV64_F28,
+    UNW_RISCV64_F29,
+    UNW_RISCV64_F30,
+    UNW_RISCV64_F31,
+
+    UNW_RISCV64_PC,
+
+    UNW_TDEP_LAST_REG = UNW_RISCV64_PC,
+
+    /* The CFA is the value of SP in previous frame */
+    UNW_RISCV64_CFA = UNW_RISCV64_X2,
+
+    UNW_TDEP_IP = UNW_RISCV64_PC,
+    UNW_TDEP_SP = UNW_RISCV64_X2,
+    UNW_TDEP_EH = UNW_RISCV64_X1, /* FIXME.  */
```

Comment:
This should be X10. Most architectures use the argument registers for this. See https://github.com/gcc-mirror/gcc/blob/16e2427f50c208dfe07d07f18009969502c25dc8/gcc/config/riscv/riscv.h#L104-L106.